File: backend/voidswitch/api/admin/tokens.py
"""Admin: manage Void-Tokens across all users."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from voidswitch.core.audit import record_audit
from voidswitch.core.auth import actor_display_name, require_owner
from voidswitch.core.database import get_session
from voidswitch.core.security import generate_void_token, hash_token, token_fingerprint
from voidswitch.models.db import User, VoidToken
from voidswitch.models.schemas import (
    VoidTokenCreate,
    VoidTokenOut,
    VoidTokenUpdate,
    VoidTokenWithSecret,
)

router = APIRouter(prefix="/api/admin/tokens", tags=["admin:tokens"])


@router.get("", response_model=list[VoidTokenOut])
async def list_tokens(
    user_id: int | None = None,
    session: AsyncSession = Depends(get_session),
    _: User = Depends(require_owner),
) -> list[VoidToken]:
    stmt = select(VoidToken).order_by(VoidToken.id)
    if user_id is not None:
        stmt = stmt.where(VoidToken.user_id == user_id)
    rows = (await session.execute(stmt)).scalars().all()
    return list(rows)


@router.post("", response_model=VoidTokenWithSecret, status_code=status.HTTP_201_CREATED)
async def create_token(
    body: VoidTokenCreate,
    request: Request,
    session: AsyncSession = Depends(get_session),
    actor: User = Depends(require_owner),
) -> VoidTokenWithSecret:
    target_user_id = body.user_id or actor.id
    owner = await session.get(User, target_user_id)
    if owner is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Target user not found.")
    secret = generate_void_token()
    token = VoidToken(
        user=owner,
        name=body.name,
        token_hash=hash_token(secret),
        token_prefix=token_fingerprint(secret),
        allowed_models=body.allowed_models,
        rpm_limit=body.rpm_limit,
        daily_quota=body.daily_quota,
        expires_at=body.expires_at,
    )
    session.add(token)
    await session.flush()
    await record_audit(
        session,
        action="token.create",
        actor_sub=actor.sub,
        actor_name=actor_display_name(actor),
        target_type="token",
        target_id=token.id,
        detail={"name": token.name, "user_id": owner.id},
        ip=request.client.host if request.client else None,
    )
    # Plaintext secret returned exactly once; ORM stores only its hash.
    return VoidTokenWithSecret(**VoidTokenOut.model_validate(token).model_dump(), token=secret)


@router.patch("/{token_id}", response_model=VoidTokenOut)
async def update_token(
    token_id: int,
    body: VoidTokenUpdate,
    request: Request,
    session: AsyncSession = Depends(get_session),
    actor: User = Depends(require_owner),
) -> VoidToken:
    token = await session.get(VoidToken, token_id)
    if token is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Token not found.")
    changes = body.model_dump(exclude_unset=True)
    for field, value in changes.items():
        setattr(token, field, value)
    await session.flush()
    await record_audit(
        session,
        action="token.update",
        actor_sub=actor.sub,
        actor_name=actor_display_name(actor),
        target_type="token",
        target_id=token.id,
        # mode="json" so datetime fields (e.g. expires_at) land as strings the
        # JSON column can store.
        detail={
            "user_id": token.user_id,
            "changes": body.model_dump(mode="json", exclude_unset=True),
        },
        ip=request.client.host if request.client else None,
    )
    return token


@router.delete("/{token_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_token(
    token_id: int,
    request: Request,
    session: AsyncSession = Depends(get_session),
    actor: User = Depends(require_owner),
) -> None:
    token = await session.get(VoidToken, token_id)
    if token is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Token not found.")
    await record_audit(
        session,
        action="token.delete",
        actor_sub=actor.sub,
        actor_name=actor_display_name(actor),
        target_type="token",
        target_id=token.id,
        detail={"name": token.name, "user_id": token.user_id},
        ip=request.client.host if request.client else None,
    )
    await session.delete(token)

File: backend-go/internal/handlers/admin/tokens.go
package admin

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"

	"github.com/siiway/voidswitch/internal/core"
	"github.com/siiway/voidswitch/internal/database"
	"github.com/siiway/voidswitch/internal/models"
)

func listTokens(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireOwnerGin(c, user) {
		return
	}

	limit := 50
	offset := 0

	if l := c.Query("limit"); l != "" {
		if parsed, err := strconv.Atoi(l); err == nil && parsed > 0 {
			limit = parsed
		}
	}
	if o := c.Query("offset"); o != "" {
		if parsed, err := strconv.Atoi(o); err == nil && parsed >= 0 {
			offset = parsed
		}
	}

	q := db.Model(&database.VoidToken{})
	userID := c.Query("user_id")
	if userID != "" {
		q = q.Where("user_id = ?", userID)
	}

	var total int64
	if err := q.Count(&total).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to count tokens."})
		return
	}

	var tokens []database.VoidToken
	if err := q.Preload("User").Limit(limit).Offset(offset).Order("id DESC").Find(&tokens).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to query tokens."})
		return
	}

	items := make([]models.VoidTokenOut, 0, len(tokens))
	for _, t := range tokens {
		items = append(items, tokenToOut(t))
	}

	c.JSON(http.StatusOK, models.Page[models.VoidTokenOut]{
		Items:  items,
		Total:  int(total),
		Limit:  limit,
		Offset: offset,
	})
}

func createToken(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireOwnerGin(c, user) {
		return
	}

	var body models.VoidTokenCreate
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body."})
		return
	}

	targetUserID := user.ID
	if body.UserID != nil && *body.UserID != 0 {
		targetUserID = *body.UserID
	}

	var targetUser database.User
	if err := db.First(&targetUser, targetUserID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Target user not found."})
		return
	}

	rawToken := core.GenerateVoidToken()
	tokenHash := core.HashToken(rawToken)
	tokenPrefix := rawToken[:len("vs-XXXXXXXX")]

	record := database.VoidToken{
		UserID:        targetUser.ID,
		Name:          body.Name,
		TokenHash:     tokenHash,
		TokenPrefix:   tokenPrefix,
		AllowedModels: body.AllowedModels,
		RpmLimit:      body.RPMLimit,
		DailyQuota:    body.DailyQuota,
		ExpiresAt:     body.ExpiresAt,
	}

	if err := db.Create(&record).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create token."})
		return
	}

	name := core.ActorDisplayName(user)
	targetID := strconv.Itoa(record.ID)
	targetType := "void_token"
	core.RecordAudit(db, "token.create", &user.Sub, &name, &targetType, &targetID,
		map[string]any{"name": record.Name, "user_id": record.UserID}, nil, nil, nil, "admin")

	out := tokenToOut(record)
	c.JSON(http.StatusOK, models.VoidTokenWithSecret{
		VoidTokenOut: out,
		Token:        rawToken,
	})
}

func updateToken(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireOwnerGin(c, user) {
		return
	}

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid token ID."})
		return
	}

	var existing database.VoidToken
	if err := db.First(&existing, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found."})
		return
	}

	var body models.VoidTokenUpdate
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body."})
		return
	}

	if body.Name != nil {
		existing.Name = *body.Name
	}
	if body.Enabled != nil {
		existing.Enabled = *body.Enabled
	}
	if body.AllowedModels != nil {
		existing.AllowedModels = *body.AllowedModels
	}
	if body.RPMLimit != nil {
		existing.RpmLimit = *body.RPMLimit
	}
	if body.DailyQuota != nil {
		existing.DailyQuota = *body.DailyQuota
	}
	if body.ExpiresAt != nil {
		existing.ExpiresAt = body.ExpiresAt
	}

	if err := db.Save(&existing).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update token."})
		return
	}

	c.JSON(http.StatusOK, tokenToOut(existing))
}

func deleteToken(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireOwnerGin(c, user) {
		return
	}

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid token ID."})
		return
	}

	var existing database.VoidToken
	if err := db.First(&existing, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found."})
		return
	}

	if err := db.Delete(&existing).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete token."})
		return
	}

	name := core.ActorDisplayName(user)
	targetID := strconv.Itoa(id)
	targetType := "void_token"
	core.RecordAudit(db, "token.delete", &user.Sub, &name, &targetType, &targetID,
		map[string]any{"name": existing.Name}, nil, nil, nil, "admin")

	c.Status(http.StatusNoContent)
}

func tokenToOut(t database.VoidToken) models.VoidTokenOut {
	var username *string
	if t.User != nil {
		username = t.User.Username
	}
	return models.VoidTokenOut{
		ID:            t.ID,
		UserID:        t.UserID,
		Username:      username,
		Name:          t.Name,
		TokenPrefix:   t.TokenPrefix,
		Enabled:       t.Enabled,
		AllowedModels: t.AllowedModels,
		RPMLimit:      t.RpmLimit,
		DailyQuota:    t.DailyQuota,
		TotalRequests: t.TotalRequests,
		TotalTokens:   t.TotalTokens,
		LastUsedAt:    t.LastUsedAt,
		ExpiresAt:     t.ExpiresAt,
		CreatedAt:     t.CreatedAt,
	}
}

func RegisterTokenRoutes(router *gin.RouterGroup) {
	tokenGroup := router.Group("/tokens")
	{
		tokenGroup.GET("", listTokens)
		tokenGroup.POST("", createToken)
		tokenGroup.PATCH("/:id", updateToken)
		tokenGroup.DELETE("/:id", deleteToken)
	}
}
